File: app/src/main/java/com/example/alertdrive/Custom_Adapter.java
package com.example.alertdrive;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

public class Custom_Adapter extends ArrayAdapter<String> {

    Context mContext;
    List<String> todo = new ArrayList<>();
    List<Boolean> compl = new ArrayList<>();


    // View lookup cache
    private static class ViewHolder {
        ExpandableTextView txtName;
        ImageView txtType;

    }

    public Custom_Adapter(List<String> todo, Context context, List<Boolean> compl) {
        super(context, R.layout.list_row, todo);
        this.todo = todo;
        this.compl = compl;
        this.mContext=context;

    }



    private int lastPosition = -1;

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        ViewHolder viewHolder; // view lookup cache stored in tag

        final View result;

        if (convertView == null) {

            viewHolder = new ViewHolder();
            LayoutInflater inflater = LayoutInflater.from(getContext());
            convertView = inflater.inflate(R.layout.list_row, parent, false);
            viewHolder.txtName = (ExpandableTextView) convertView.findViewById(R.id.one);
            viewHolder.txtType = convertView.findViewById(R.id.text1);



            result=convertView;

            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
            result=convertView;
        }

        lastPosition = position;

        viewHolder.txtName.setText(todo.get(position));
        if(compl.get(position)) {
            viewHolder.txtType.setVisibility(View.VISIBLE);
        }else{
            viewHolder.txtType.setVisibility(View.INVISIBLE);
        }

        // Return the completed view to render on screen
        return convertView;
    }

    public void update(List<String> list, List<Boolean> checkList){
        todo.clear();
        compl.clear();
        todo.addAll(list);
        compl.addAll(checkList);
        this.notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/alertdrive/Custom_Adapter.kt
package com.example.alertdrive

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import java.util.ArrayList

class Custom_Adapter(todo: MutableList<String>, private val mContext: Context, compl: MutableList<Boolean>) : ArrayAdapter<String>(mContext, R.layout.list_row, todo) {
    private var todo = ArrayList<String>()
    private var compl = ArrayList<Boolean>()


    private class ViewHolder {
        internal var txtName: ExpandableTextView? = null
        internal var txtType: ImageView? = null
    }

    init {
        this.todo = todo as ArrayList<String>
        this.compl = compl as ArrayList<Boolean>
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView

        val viewHolder: ViewHolder


        if (convertView == null) {
            viewHolder = ViewHolder()
            val inflater = LayoutInflater.from(context)
            convertView = inflater.inflate(R.layout.list_row, parent, false)
            viewHolder.txtName = convertView!!.findViewById<View>(R.id.one) as ExpandableTextView
            viewHolder.txtType = convertView.findViewById(R.id.text1)
            convertView.tag = viewHolder
        } else {
            viewHolder = convertView.tag as ViewHolder
        }
        viewHolder.txtName!!.text = todo[position]
        if (compl[position]) {
            viewHolder.txtType!!.visibility = View.VISIBLE
        } else {
            viewHolder.txtType!!.visibility = View.INVISIBLE
        }
        return convertView
    }

    fun update(list: List<String>, checkList: List<Boolean>) {
        todo.clear()
        compl.clear()
        todo.addAll(list)
        compl.addAll(checkList)
        this.notifyDataSetChanged()
    }
}